spatialprofilingtoolbox/ondemand/request_handler.py

```diff
@@ -0,0 +1,186 @@
+"""Handler for requests for on demand calculations."""
+
+from socketserver import BaseRequestHandler
+from json import dumps
+
+from spatialprofilingtoolbox.db.exchange_data_formats.metrics import PhenotypeCriteria
+from spatialprofilingtoolbox.ondemand.tcp_server import OnDemandTCPServer
+from spatialprofilingtoolbox.standalone_utilities.log_formats import colorized_logger
+
+logger = colorized_logger('spt ondemand start')
+
+
+class OnDemandRequestHandler(BaseRequestHandler):
+    """TCP server for on demand metrics."""
+
+    server: OnDemandTCPServer
+
+    def handle(self):
+        """Handle an on demand request."""
+        data = self.request.recv(512).strip()
+        logger.info('Request: %s', data)
+        if self._handle_empty_body(data):
+            return
+        if self._handle_single_phenotype_counts_request(data):
+            return
+        if self._handle_proximity_request(data):
+            return
+        if self._handle_squidpy_request(data):
+            return
+
+    def _handle_proximity_request(self, data):
+        groups = self._get_groups(data)
+        if len(groups) != 6:
+            return False
+        specification = self._get_phenotype_pair_specification(groups)
+        study_name = specification[0]
+        if self._handle_missing_study(study_name):
+            return True
+
+        try:
+            metrics = self._get_proximity_metrics(*specification)
+        except Exception as exception:
+            message = "Error response." + self._get_end_of_transmission()
+            self.request.sendall(message.encode('utf-8'))
+            raise exception
+
+        message = dumps(metrics) + self._get_end_of_transmission()
+        self.request.sendall(message.encode('utf-8'))
+        return True
+
+    def _get_proximity_metrics(self, study, radius, signature):
+        positives1, negatives1, positives2, negatives2 = signature
+        phenotype1 = PhenotypeCriteria(positives1, negatives1)
+        phenotype2 = PhenotypeCriteria(positives2, negatives2)
```

Comment:
One of the downsides of Pydantic as compared with `attrs` is that this constructor format is not supported. In Pydantic you have to provide the controlled attributes by keyword.

Reply:
We know how to keyword type now so that's less of an issue.